File: 2017/d05.py
with open("in/d05.txt") as f:
    data = tuple(int(x.strip()) for x in f)


def solver(xs: list[int], func=lambda _: 1) -> int:
    steps = 0
    i = 0
    while 0 <= i < len(xs):
        jump = xs[i]
        xs[i] += func(jump)
        i += jump
        steps += 1
    return steps


print(f"Part 1: {solver(list(data))}")
print(f"Part 2: {solver(list(data), lambda x: -1 if x >= 3 else 1)}")

File: 2018/aoc/d05.rs
use std::fs;

fn process_data(mut data: Vec<char>) -> usize {
    let mut tmp: Vec<char> = vec![];

    loop {
        tmp.clear();
        let mut i = 0 as usize;

        while i < data.len() {
            if i + 1 != data.len() {
                if data[i].to_ascii_lowercase() == data[i + 1].to_ascii_lowercase() {
                    if data[i] != data[i + 1] {
                        i += 2;
                        continue;
                    }
                }
            }
            tmp.push(data[i]);
            i += 1;
        }
        let last_size = data.len();
        data = tmp.clone();
        if tmp.len() == last_size {
            return tmp.len();
        }
    }
}

fn main() -> std::io::Result<()> {
    let contents: Vec<char> = fs::read_to_string("in/d05.txt")?
        .trim_end()
        .chars()
        .collect();

    let p2_res = ('a'..='z')
        .map(|c| {
            let data: Vec<char> = contents
                .iter()
                .filter(|v| v.to_ascii_lowercase() != c)
                .cloned()
                .collect();
            process_data(data)
        })
        .min()
        .unwrap();

    println!("Part 1: {}", process_data(contents));
    println!("Part 2: {}", p2_res);
    Ok(())
}
